cpp/src/arrow/util/bit_util.h

```diff
@@ -167,10 +167,8 @@ constexpr int64_t CoveringBytes(int64_t offset, int64_t length) {
 
 // Returns the 'num_bits' least-significant bits of 'v'.
 static inline uint64_t TrailingBits(uint64_t v, int num_bits) {
-  if (ARROW_PREDICT_FALSE(num_bits == 0)) return 0;
```

Comment:
(so generally, checks `num_bits == 0` is not highly related to the optimization?

Reply:
yes, because ((v >> 0) << 0) ^ v == v ^ v == 0

Reply:
and i think that is the main reason for performance increase

Reply:
> and i think that is the main reason for performance increase

As I benchmark. Without `-march=skylake`, removing the `ARROW_PREDICT_FALSE(num_bits == 0)` would not make it faster or slower.  And as mentioned https://github.com/apache/arrow/pull/39403#issuecomment-1873378808  . It might affect the command generation later. Would you mind testing that?